components/Workspaces/AddWorkspaceModal/index.tsx

```diff
@@ -0,0 +1,141 @@
+import {
+  Modal,
+  ModalBody,
+  ModalCloseButton,
+  ModalContent,
+  ModalHeader,
+  ModalOverlay,
+  FormControl,
+  FormErrorMessage,
+  FormLabel,
+  Input,
+  ModalFooter,
+  Button,
+} from '@chakra-ui/react';
+import * as Yup from 'yup';
+import { useFormik } from 'formik';
+import { useRef } from 'react';
+import { useMutation } from '@apollo/client';
+import { v4 as uuidv4 } from 'uuid';
+import { UserQueryResponse, USER_QUERY } from '@/queries/userQuery';
+import {
+  CreateWorkspaceMutationResponse,
+  CREATE_WORKSPACE_MUTATION,
+} from '@/mutations/createWorkspace';
+import { useRouter } from 'next/router';
+
+interface Props {
+  isOpen: boolean;
+  onClose: () => void;
+}
+
+const validationSchema = Yup.object().shape({
+  workspaceName: Yup.string().required('Required'),
+});
+
+const AddWorkspaceModal: React.FC<Props> = ({ isOpen, onClose }) => {
+  const [mutate] = useMutation<CreateWorkspaceMutationResponse>(
+    CREATE_WORKSPACE_MUTATION,
+  );
+  const router = useRouter();
+  const {
+    errors,
+    touched,
+    values,
+    isValid,
+    isSubmitting,
+    handleChange,
+    handleSubmit,
+  } = useFormik({
+    initialValues: {
+      workspaceName: '',
+    },
+    initialErrors: {
+      workspaceName: '',
+    },
+    validationSchema,
+    onSubmit: async ({ workspaceName }) => {
+      const result = await mutate({
+        variables: { name: workspaceName },
+        optimisticResponse: {
+          createWorkspace: {
+            id: uuidv4(),
+            workspaceName,
+            users: [],
+            formCount: 0,
+          },
+        },
+        update: (proxy, { data: { createWorkspace } }) => {
+          const cache = proxy.readQuery<UserQueryResponse>({
+            query: USER_QUERY,
+          });
+
+          proxy.writeQuery({
+            query: USER_QUERY,
+            data: {
+              user: {
+                ...cache.user,
```

Comment:
no clue what this is doing

Reply:
It allows us to update the UI before we actually have a response from the network. Optimistic response pattern from apollo